### PR TITLE
Added HystrixGroupKey annotation

### LINKED ISSUE #334: allow anotation of hystrix commandKey and groupKey for interfaced feign methods
There are defaults for cmdKesy = methodname and groupKey=classname.
 If hytsrix cfg in application.yml is linked via this default names and method or class names get refactored the configuration is delinked without recognizing. Therefore explict labeling via custom commandKey and groupKey anotated above the Feign (interface) method would help to explicit link that and thre are no problems with code refactorings.

This is related to https://github.com/Netflix/feign/issues/189#issuecomment-95688652

### PR DESCRIPTION
**Pull request by caarlos0**
Added `@HystrixConfig` and `@HystrixCommandConfig` annotations so we can set a custom group and command key for each service, as well timeouts. We could add more config in the future.

If none is set, defaults to the previous behavior.

Syntax will be like:

```java
    @HystrixConfig(key = "MyKey", timeout = 150)
    interface Service {
        @RequestLine("GET /status")
        HystrixCommand<String> status();

        @HystrixCommandConfig(key = "Slow", timeout = 15000)
        @RequestLine("GET /slower")
        HystrixCommand<String> slower();
    }
```

this should also fix #334

### CODE REVIEW

**Comment by caarlos0** [comment]
ping?

**Comment by adriancole** [comment]
sorry, I meant to comment. For example, what about the command key? For example, javanica annotations looks like this: 

`@HystrixCommand(groupKey="UserGroup", commandKey = "GetUserByIdCommand")`

I think there are others who've asked about setting hystrix group key. Can you look through similar issues and ping them on this issue? I'd like this type of feature to be diversely supported.

**Comment by caarlos0** [comment]
@adriancole makes sense, I guess I can join all configs in one single annotation. Will do that and ping you back (and find other issues related).
Thanks :dancer:

**Comment by adriancole** [comment]
ps for an implementation I'd just start with renaming the annotation and adding groupKey, commandKey.

However, this is kiting a bit, assuming that this is a good way to go. can you ping others who have contributed to hystrix here in the past, like @spencergibb, to see if they agree with this path?

**Comment by caarlos0** [comment]
@adriancole I did it, but I think I found some thread-safety issues, wasn't able to find where though... can you give me some advice about it?

File: hystrix/src/main/java/feign/hystrix/HystrixCommandConfig.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ *     }
+ * </pre>
+ * </blockquote>
+ */
+@Target(METHOD)
+@Retention(RUNTIME)
+public @interface HystrixCommandConfig {
+    String key() default "";
+
+    int timeout() default -1;
```
Comment (adriancole) [inline hystrix/src/main/java/feign/hystrix/HystrixCommandConfig.java]: I've fallen down this trap before, and wouldn't recommend doing it again. timeout values often have far more to do with the network than the api, and this annotation puts this in code vs config.

File: hystrix/src/main/java/feign/hystrix/HystrixConfig.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ *     @HystrixConfig(key = "MyCustomKey", timeout = 2000)
+ *     interface MyService {
+ *         // methods...
+ *     }
+ * </pre>
+ * </blockquote>
+ */
+@Target(TYPE)
+@Retention(RUNTIME)
+public @interface HystrixConfig {
```
Comment (adriancole) [inline hystrix/src/main/java/feign/hystrix/HystrixConfig.java]: I don't want to support multiple annotations at this point.. can we choose one?

File: hystrix/src/main/java/feign/hystrix/HystrixInvocationHandler.java [hunk 3]
```diff
@@ -106,6 +102,34 @@ protected Object getFallback() {
     return hystrixCommand.execute();
   }
 
+  private HystrixCommand.Setter config(final Object proxy, final Method method) {
```
Comment (adriancole) [inline hystrix/src/main/java/feign/hystrix/HystrixInvocationHandler.java]: probably what we want to do, is make something like MethodMetadata for Hystrix. That way, parsing annotations is separate from using them. For example, we don't want to replicate javanica. We can have a simple way to specify command and group names, and leave any further settings to either out-of-band configuration or another annotation system.

File: hystrix/src/main/java/feign/hystrix/HystrixCommandConfig.java (hunk 1 above)
Comment (caarlos0) [inline hystrix/src/main/java/feign/hystrix/HystrixCommandConfig.java]: makes sense, so, how can I change the timeout of a hystrix command if it was already "built" by feign?

File: hystrix/src/main/java/feign/hystrix/HystrixConfig.java (hunk 2 above)
Comment (caarlos0) [inline hystrix/src/main/java/feign/hystrix/HystrixConfig.java]: I guess we can...

File: hystrix/src/main/java/feign/hystrix/HystrixInvocationHandler.java (hunk 3 above)
Comment (caarlos0) [inline hystrix/src/main/java/feign/hystrix/HystrixInvocationHandler.java]: I'm not familiar with javanica..

Not sure I understand your idea... can you elaborate?

**Comment by apischan** [comment]
Whether this pull request is dead?

**Comment by adriancole** [comment]
/me pokes with a stick

Seems like it. We should generally close dead PRs since they distract folks.

**Comment by caarlos0** [comment]
Yeah, I was kind of waiting for a reply here: https://github.com/OpenFeign/feign/pull/360#discussion_r58198701

**Comment by caarlos0** [comment]
but, np, will close it.

**Comment by adriancole** [comment]
I meant something like this:
https://github.com/OpenFeign/feign/issues/434#issuecomment-239609848

**Pull request closed without merging**